exercises/slot_machine.livemd

```diff
@@ -46,46 +49,91 @@ defmodule SlotMachine do
   Make sure that the machine doesn't give more rewards than its balance.
   Make sure the machine doesn't give negative rewards!
 
-  Use the :rand erlang module to generate random numbers.
-
-  The machine should either print "You lose!" or "You won ... dollars"
+  The machine should either print "You lose!" or "You won ... pennies"
   """
   def roll(pid) do
+    chance = Enum.random([1, 2, 3])
+
+    # Only give reward 1 in 3 times 
+    # Don't give reward if the machine does not have any money in it
+    if chance != 1 || get_current_balance(pid) == 0 do
+      IO.puts("You lose!")
+    else
+      # Provide random reward between 1 and 7000 pennies
+      reward = Enum.random(1..7_000)
+
+      # Don't give reward higher tham machine balance
+      if reward > get_current_balance(pid) do
+        reward = get_current_balance(pid)
+        subtract_from_balance(pid, reward)
+        IO.puts("You win #{reward} pennies!")
+      else
+        subtract_from_balance(pid, reward)
+        IO.puts("You won #{reward} pennies!")
+      end
+    end
+  end
+
+  def get_current_balance(pid) do
+    GenServer.call(pid, :get_balance)
+  end
+
+  def subtract_from_balance(pid, amount) do
+    GenServer.call(pid, {:subtract_balance, amount})
+  end
+
+  def handle_call({:subtract_balance, amount}, _from, state) do
+    state = %{balance: state.balance - amount}
+    {:reply, state, state}
+  end
+
+  def handle_call(:get_balance, _from, state) do
+    {:reply, state.balance, state}
   end
+end
+
+```
+
+</details>
+
+```elixir
+defmodule SlotMachine do
+  @moduledoc """
+  The slot machine GenServer
+  """
+  use GenServer
 
   @doc """
-  Get the current slot machine balance.
+    Start the slot machine
 
-  ## Example
+    ## Example
 
-    iex> {:ok, pid} = SlotMachine.start_link()
-    iex> GenServer.call(pid, {:get_balance})
-    7000
+      iex> SlotMachine.start_link(1)
+      {:ok, pid}
 
   """
-  def handle_call({:get_balance}, _from, state) do
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, :ok)
   end
 
-  @doc """
-  Subtract pennies from slot machine balance.
+  def init(_) do
+    {:ok, %{balance: 7000}}
+  end
 
-  ## Example
+  @doc """
+  This is the main function for this module.
 
-    iex> {:ok, pid} = SlotMachine.start_link()
-    iex> GenServer.call(pid, {:subtract_balance, 1000})
-    6000
+  Make sure that that players get rewards only 30% of the time.
```

Comment:
1/3 is not `30%`. You'd either need to change this description to 1 in 3 ods or change the random function to something like `winner = Enum.random(1..100) <= 30`